qutip/nonmarkov/transfertensor.py

```diff
@@ -148,10 +147,12 @@ def ttmsolve(dynmaps, rho0, times, e_ops=[], learningtimes=None, tensors=None,
     K = len(tensors)
     states = [rho0vec]
     for n in range(1, len(times)):
+        # Append empty state
         states.append(None)
         for k in range(n):
-            if n-k < K:
-                states[-1] += tensors[n-k]*states[k]
+            if n - k < K:
+                tmp = tensors[n - k] * states[k]
+                states[-1] = tmp if states[-1] is None else tmp + states[-1]
```

Comment:
This whole loop seems a bit iffy. `k` runs from `max(0, n - K + 1)` to `n - 1` and `n - k` runs from `K - 1` to `1`. So there are `K` tensors but only `K - 1` of them are used which is strange.

It also seems more natural to express this as `for j in range(0, min(K, n)):` with `k = n - j - 1` which would have made the missing `k` more obvious and remove the need for the tangled indexing and the `if n -  k < K`.

On the other hand the test seems to pass so maybe one of the tensors is meant to be ignored? Do you perhaps know what is supposed to happen here?

Reply:
I would also just set `state = None` outside the loop and do `state = tmp if state is None else tmp + state` and then do `states.append(state)` afterwards (this also remove some awkward `-1` terms from the indexing.

Reply:
I don't know why the first tensor is left out. In order to drop the `if n-k < K` expression I changed the `k` for loop by defining `j = n - k`. This resolves to a for loop `for j in range(1, min(K, n+1))`.

And I replaced the `states.append(None)` part as you described.